### DISCUSSION TITLE
Distilling VIT-L from VIT-G on imagenet

### DISCUSSION THREAD

**Discussion by nighting0le01** [yoshitomo-matsubara/torchdistill#374] [Q&A] [answered]
Hi, How can modify the config files to Distill from a ViT-g model pretrained on Imagenet21k& finetuned on Imagenet 1k to a VIT-L model. 
What All do I need to modify
both checkpoints are avaialble here https://github.com/huggingface/pytorch-image-models

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @arpitsahni04 

Do you just want to change models in some ImageNet configs?
If so, with the current PyPI package (torchdistill==0.3.3), your model entry in a yaml config will look like this as shown in [README.md](https://github.com/yoshitomo-matsubara/torchdistill#pytorch-hub)

```
models:
  teacher_model:
    name: 'vit_giant_patch14_224'
    repo_or_dir: 'huggingface/pytorch-image-models'
    params:
      pretrained: True
      other kwargs here if necessary ...
  student_model:
    name: 'vit_large_patch14_clip_224'
    repo_or_dir: 'huggingface/pytorch-image-models'
    params:
      pretrained: False
      other kwargs here if necessary ...
```

For the teacher and student names, I referred to https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/vision_transformer.py

Note that if you'll need to tune hyperparameters if you reuse existing yaml config files under `sample/` like [this](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/legacy/sample/ilsvrc2012/single_stage/kd/resnet18_from_resnet152.yaml)

**Reply by nighting0le01** [reply]
Hi @yoshitomo-matsubara yes I want to run experiments on ImageNet 1k with vit-g and vit-L: 


the sample file you referenced in your post doesn't have repo_or_dir key. also do i need to edit the config and give it to this:[image_classification.py](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/examples/legacy/image_classification.py)


```
models:
  teacher_model:
    name: &teacher_model_name 'resnet152'
    params:
      num_classes: 1000
      pretrained: True
    experiment: &teacher_experiment !join [*dataset_name, '-', *teacher_model_name]
    ckpt: !join ['./resource/ckpt/ilsvrc2012/teacher/', *teacher_experiment, '.pt']
  student_model:
    name: &student_model_name 'resnet18'
    params:
      num_classes: 1000
      pretrained: False
    experiment: &student_experiment !join [*dataset_name, '-', *student_model_name, '_from_', *teacher_model_name]
    ckpt: !join ['./resource/ckpt/ilsvrc2012/single_stage/kd/', *student_experiment, '.pt']
```
2.  how do I just evalute a distilled model on imagenet validation set.

**Reply by nighting0le01** [reply]
Also @yoshitomo-matsubara 
where can i get the right hyperparameters to compare 
google/vit-large-patch16-224
finetuned via distillation on imagenet 1k with google/vit-large-patch16-224-in21k

**Reply by yoshitomo-matsubara** [reply]
Hi @arpitsahni04 
1. You replace whole `models` entry with one I provied above, and you give the yaml config to image_classification.py with `--config` 
2. If you want to skip training, add `-test_only` and `-student_only` 
3. Check `timm` documentation and its original papers for hyperparameters they used. I've never used these specific models

**Reply by nighting0le01** [reply]
Thanks @yoshitomo-matsubara  alot what are other kwargs? can you share the entire template for it so i can go through it and give the right parameters?
```

models:
  teacher_model:
    name: 'vit_giant_patch14_224'
    repo_or_dir: 'huggingface/pytorch-image-models'
    params:
      pretrained: True
      other kwargs here if necessary ...
  student_model:
    name: 'vit_large_patch14_clip_224'
    repo_or_dir: 'huggingface/pytorch-image-models'
    params:
      pretrained: False
      other kwargs here if necessary ...
```

Also how do i use models from hugging face hub: example https://huggingface.co/google/vit-large-patch16-224-in21k or https://huggingface.co/google/vit-large-patch16-224 what repo_or_dir and name do i need

**Reply by yoshitomo-matsubara** [reply]
Check `timm`'s documentation for kwargs of `vit_giant_patch14_224` and `vit_large_patch14_clip_224` functions as you're importing models in `timm` via PyTorch Hub. If you don't specify them in the yaml config file, their default parameters will be used.

Also I've never tried to import models on HuggingFace from image_classification.py, so I have no ideas

**Reply by nighting0le01** [reply]
where can i find the list of models that have:
1. vit-g : pretrained on img_net 21k and ft on img_net 1k
2. vit- L pretrained on img_net 21k
3. vit-L pretrained on img_net21k and finetuned on img_net1k.
does the image_classification.py only support config that have models imported from timm? can we use models like this https://huggingface.co/google/vit-large-patch16-224-in21k in it??

**Reply by yoshitomo-matsubara** [reply]
torchdistill v0.3.3 does not directly support timm, but instead supports torch.hub and torchvision. If you want to use models in torchvision you installed to your machine, you can use it in a way many sample configs show.

In your case, the timm supports torch hub and you can import the models using torch hub, and you need to check how to call the model functions in timm by yourself.

If you want to use your own models, check other existing discussions like [this](https://github.com/yoshitomo-matsubara/torchdistill/discussions/278)

**Reply by yoshitomo-matsubara** [reply]
Please keep the discussion in one thread

I don't quite understand what you meant by the pretraining and finetuning requirements. I just found random vit_giant and vit_large model functions from https://github.com/huggingface/pytorch-image-models as teacher and student models respectively (teacher is `pretrained: True` to use timm's model weights and student is `pretrained: False` to train scratch. The `pretrained` is a part of their kwargs, not of torchdistill)

**Reply by nighting0le01** [reply]
Thanks, 
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/81643693/e52a666a-c55f-4a04-a697-c05ff8d24b15)

![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/81643693/409f01db-2883-4f0f-a904-d385fb664239)
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/81643693/0b3990bf-9276-4cfa-a88f-ef001587387d)
1. why am i getting this key error?
2. can you give a example of a config model with a student and teacehr vit that works in torchdistill?

**Reply by yoshitomo-matsubara** [reply]
See the error message. You replaced `params` in the yaml I suggested above with `kwargs`, which will be effective in the next major release but not valid for the current version (v0.3.3).

Click "Mark as answer" if it answers your initial question.

**Reply by nighting0le01** [reply]
Oh I see thanks! Would it be possible to please give a example config with VITs? There are none like that in the examples. Thanks really appreciate it

**Reply by yoshitomo-matsubara** [reply]
I am currently busy with reviewing papers at multiple venues and do not have a buffer to create and test a new sample config for you

**Reply by nighting0le01** [reply]
can you please give any. direction to create it ? do we need to specify ckpt also? you hadn't specified it
'''
models:
  teacher_model:
    name: 'vit_giant_patch14_224'
    repo_or_dir: 'huggingface/pytorch-image-models'
    params:
      pretrained: True
      other kwargs here if necessary ...
  student_model:
    name: 'vit_large_patch14_clip_224'
    repo_or_dir: 'huggingface/pytorch-image-models'
    params:
      pretrained: False
      other kwargs here if necessary ...

'''

**Reply by yoshitomo-matsubara** [reply]
ckpt can be empty "ckpt:" unless you have your own checkpoint file

**Reply by nighting0le01** [reply]
hi can you please explain what you mean here? so for using timm models do i need to modify load_model function for using models from timm?
```In your case, the timm supports torch hub and you can import the models using torch hub, and you need to check how to call the model functions in timm by yourself.```

![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/81643693/2d242f8f-dfdf-43e2-b786-7a291925c7fe)

**Reply by yoshitomo-matsubara** [reply]
If you want to use models in other packages like `timm` directly (i.e., without use of torch hub), you will need to edit `image_classification.py` 
Otherwise, I think you can use the code as is because the if statement should be called with the yaml config shared above.

**Reply by nighting0le01** [reply]
wait now i'm confused 
aren't the models you mentioned from timm? the ones you mentioned in the initial reply and gave example of using directly in the config? 
do i need to add a additional preprocessor to use these models? or their augmentations which are transformer specific or are they taken care of? 
```
For the teacher and student names, I referred to https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/vision_transformer.py
```
what changes do i need to make to use the models & ckpts from timm : https://github.com/rwightman/pytorch-image-models

**Reply by yoshitomo-matsubara** [reply]
Read the thread carefully.  I clearly said
> torchdistill v0.3.3 does not directly support timm, but instead supports torch.hub and torchvision. If you want to use models in torchvision you installed to your machine, you can use it in a way many sample configs show.

`resnet*` specified in the sample configs refer to the functions in torchvision.
You may need to change other entries like `transform_params` in the yaml config for `timm` models. Please check their repo what transforms they used for training and validation, and make changes in the config accordingly. You need to define almost all the components in one yaml config. This is also a key concept of torchdistill.

I also strongly suggest you edit and play with yaml configs by yourself. You should also be familiar with modules in PyTorch and PyYAML. I cannot explain everything from scratch.

**Reply by nighting0le01** [reply]
Thanks @yoshitomo-matsubara I think i've understood what to do. Just 1 thing. Does torchdistil apply the transformations listed in the default_config that we get from a model from torchhub? or do we need to take those transformations and specify them manually in the config? 
```
import torch.hub
namespace = 'rwightman/pytorch-image-models'
available_models = torch.hub.list(namespace)
model_name = 'eva_giant_patch14_224'
if model_name in available_models:
    print(f"{model_name} is available in the {namespace} repository.")
    model = torch.hub.load(namespace, model_name)
    print(model.default_cfg)
else:
    print(f"{model_name} is not available in the {namespace} repository.")
```
TEACHER:
default config returned: {'url': '', 'hf_hub_id': 'timm/eva_giant_patch14_224.clip_ft_in1k', 'architecture': 'eva_giant_patch14_224', 'tag': 'clip_ft_in1k', 'custom_load': False, 'input_size': (3, 224, 224), 'fixed_input_size': True, 'interpolation': 'bicubic', 'crop_pct': 0.9, 'crop_mode': 'center', 'mean': (0.48145466, 0.4578275, 0.40821073), 'std': (0.26862954, 0.26130258, 0.27577711), 'num_classes': 1000, 'pool_size': None, 'first_conv': 'patch_embed.proj', 'classifier': 'head', 'license': 'mit'}

STUDENT:
{'url': 'https://storage.googleapis.com/vit_models/augreg/L_16-i21k-300ep-lr_0.001-aug_medium1-wd_0.1-do_0.1-sd_0.1.npz', 'hf_hub_id': 'timm/vit_large_patch16_224.augreg_in21k', 'architecture': 'vit_large_patch16_224', 'tag': 'augreg_in21k', 'custom_load': True, 'input_size': (3, 224, 224), 'fixed_input_size': True, 'interpolation': 'bicubic', 'crop_pct': 0.9, 'crop_mode': 'center', 'mean': (0.5, 0.5, 0.5), 'std': (0.5, 0.5, 0.5), 'num_classes': 21843, 'pool_size': None, 'first_conv': 'patch_embed.proj', 'classifier': 'head'}



----also what to do when teacher and student require different transformations?? In my case the mean and sd is different as you can see

**Reply by yoshitomo-matsubara** [reply]
You will need to replace the sequence of transforms in the yaml file (`transform_params`) with those used for both **teacher** and **student** models you found (See `timm`'s implementation of the transform).

And, you should choose a different pair of teacher and student models as those should share the same transforms because dataset classes in torchvision can have only one transform for a batch of images.

**Reply by nighting0le01** [reply]
Thanks @yoshitomo-matsubara  but would using teachers transform be good enough since i'm finetuning student on imagenet1k during distillation anyway? I know wrong transformation gives out garbage so can you suggest any way to modify the image_classification.py to apply model dependent transforms

**Reply by yoshitomo-matsubara** [reply]
If your student model is trained from scratch, it may be ok to use teacher's transform. I feel that it's not easy for the torchdistill v0.3.3 to support two separate batches for teacher and student models.

**Reply by nighting0le01** [reply]
would it be feasible to apply the student and teacher transforms after the data comes out of the dataloader for student and teacher? can i make this change easily in torchdistill?

**Reply by yoshitomo-matsubara** [reply]
If your teacher and student models have their own transforms inside, it could be. But it's still not easy, and I don't know if it's worth spending time. I'd try to find a different pair of teacher and student models using the same transform

**Reply by nighting0le01** [reply]
i see @yoshitomo-matsubara thanks. but this can be the case for transformer student teacher models having different pretraining and finetuning. in my case student is pretrained on in_21k and teacher on liaon5b and fine tuned on in_1k. I'm using in_1k to distill knowledge into student. essentially distilling while finetuning after pretraining.
